protocol/msgpack/unaggregated_iterator.go

```diff
@@ -219,37 +222,62 @@ func (it *unaggregatedIterator) decodeGauge() {
 }
 
-func (it *unaggregatedIterator) decodeVersionedPolicies() {
+func (it *unaggregatedIterator) decodePoliciesList() {
 	numActualFields := it.decodeNumObjectFields()
-	versionedPoliciesType := it.decodeObjectType()
-	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
-		versionedPoliciesType,
+	policiesListType := it.decodeObjectType()
+	numExpectedFields, ok := it.checkExpectedNumFieldsForType(
+		policiesListType,
 		numActualFields,
 	)
 	if !ok {
 		return
 	}
-	version := it.decodeVersion()
-	cutover := it.decodeTime()
-	switch versionedPoliciesType {
-	case defaultVersionedPoliciesType:
-		it.versionedPolicies = policy.DefaultVersionedPolicies(version, cutover)
-		it.skip(numActualFields - numExpectedFields)
-	case customVersionedPoliciesType:
-		numPolicies := it.decodeArrayLen()
-		if cap(it.policies) < numPolicies {
-			it.policies = make([]policy.Policy, 0, numPolicies)
+	switch policiesListType {
+	case defaultPoliciesListType:
+		it.policiesList = policy.DefaultPoliciesList
+	case customPoliciesListType:
+		numStagedPolicies := it.decodeArrayLen()
+		if cap(it.cachedPoliciesList) < numStagedPolicies {
+			it.cachedPoliciesList = make(policy.PoliciesList, 0, numStagedPolicies)
 		} else {
-			it.policies = it.policies[:0]
+			it.cachedPoliciesList = it.cachedPoliciesList[:0]
+		}
+		if len(it.cachedPolicies) < numStagedPolicies {
+			it.cachedPolicies = make([][]policy.Policy, numStagedPolicies)
 		}
-		for i := 0; i < numPolicies; i++ {
-			it.policies = append(it.policies, it.decodePolicy())
+		for policyIdx := 0; policyIdx < numStagedPolicies; policyIdx++ {
+			decodedStagedPolicies := it.decodeStagedPolicies(policyIdx)
+			it.cachedPoliciesList = append(it.cachedPoliciesList, decodedStagedPolicies)
 		}
-		it.versionedPolicies = policy.CustomVersionedPolicies(version, cutover, it.policies)
-		it.skip(numActualFields - numExpectedFields)
+		it.policiesList = it.cachedPoliciesList
 	default:
-		it.setErr(fmt.Errorf("unrecognized versioned policies type: %v", versionedPoliciesType))
+		it.setErr(fmt.Errorf("unrecognized policies list type: %v", policiesListType))
 	}
+	it.skip(numActualFields - numExpectedFields)
+}
+
+// decodeStagedPolicies decodes a staged policies using the cached policies
+// to avoid memory reallocations and returns the decoded staged policies.
+// If an error is encountered during decoding, it is stored in the iterator.
+func (it *unaggregatedIterator) decodeStagedPolicies(policyIdx int) policy.StagedPolicies {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(stagedPoliciesType)
+	if !ok {
+		return policy.DefaultStagedPolicies
+	}
+	cutoverNanos := it.decodeVarint()
+	tombstoned := it.decodeBool()
+	numPolicies := it.decodeArrayLen()
+	if cap(it.cachedPolicies[policyIdx]) < numPolicies {
+		it.cachedPolicies[policyIdx] = make([]policy.Policy, 0, numPolicies)
+	} else {
+		it.cachedPolicies[policyIdx] = it.cachedPolicies[policyIdx][:0]
```

Comment:
The elements of `it.cachedPolicies` are kept around instead of getting zeroed out so we can reuse the allocated memory for policies.